x-pack/plugins/apm/public/components/app/correlations/use_failed_transactions_correlations.ts

```diff
@@ -0,0 +1,257 @@
+      errorHistogram: undefined,
+      fieldStats: undefined,
+    });
+    setResponse.flush();
+
+    try {
+      // `responseUpdate` will be enriched with additional data with subsequent
+      // calls to the overall histogram, field candidates, field value pairs, correlation results
+      // and histogram data for statistically significant results.
+      const responseUpdate: FailedTransactionsCorrelationsResponse = {
+        ccsWarning: false,
+      };
+
+      const [overallHistogramResponse, errorHistogramRespone] =
+        await Promise.all([
+          // Initial call to fetch the overall distribution for the log-log plot.
+          callApmApi({
+            endpoint: 'POST /internal/apm/latency/overall_distribution',
+            signal: abortCtrl.current.signal,
+            params: {
+              body: {
+                ...fetchParams,
+                percentileThreshold: DEFAULT_PERCENTILE_THRESHOLD,
+              },
+            },
+          }),
+          callApmApi({
+            endpoint: 'POST /internal/apm/latency/overall_distribution',
+            signal: abortCtrl.current.signal,
+            params: {
+              body: {
+                ...fetchParams,
+                percentileThreshold: DEFAULT_PERCENTILE_THRESHOLD,
+                termFilters: [
+                  {
+                    fieldName: EVENT_OUTCOME,
+                    fieldValue: EventOutcome.failure,
+                  },
+                ],
+              },
+            },
+          }),
+        ]);
+
+      const { overallHistogram, percentileThresholdValue } =
+        overallHistogramResponse;
+      const { overallHistogram: errorHistogram } = errorHistogramRespone;
+
+      responseUpdate.errorHistogram = errorHistogram;
+      responseUpdate.overallHistogram = overallHistogram;
+      responseUpdate.percentileThresholdValue = percentileThresholdValue;
+
+      if (abortCtrl.current.signal.aborted) {
+        return;
+      }
+
+      setResponse({
+        ...responseUpdate,
+        loaded: LOADED_OVERALL_HISTOGRAM,
+      });
+      setResponse.flush();
+
+      const { fieldCandidates: candidates } = await callApmApi({
+        endpoint: 'GET /internal/apm/correlations/field_candidates',
+        signal: abortCtrl.current.signal,
+        params: {
+          query: fetchParams,
+        },
+      });
+
+      if (abortCtrl.current.signal.aborted) {
+        return;
+      }
+
+      const fieldCandidates = candidates.filter((t) => !(t === EVENT_OUTCOME));
+
+      setResponse({
+        loaded: LOADED_FIELD_CANDIDATES,
+      });
+      setResponse.flush();
+
+      const failedTransactionsCorrelations: FailedTransactionsCorrelation[] =
+        [];
+      const fieldsToSample = new Set<string>();
+      const chunkSize = 10;
+      let chunkLoadCounter = 0;
+
+      const fieldCandidatesChunks = chunk(fieldCandidates, chunkSize);
+
+      for (const fieldCandidatesChunk of fieldCandidatesChunks) {
```

Comment:
This too?

Reply:
This splits all field candidates into chunks, the chunks are called in sequence here on the client side, but all field candidates of a chunk are then queried in parallel on the Kibana server side.

Reply:
Yes, but why call these in sequence? how many blocking calls can we expect here?

Reply:
This was made in the spirit of "make it slow". I'm sure this can be further optimized, in this PR we started to parallelize the server side calls and play it safe on the client side. Since the field candidates and field value pairs are generated dynamically, we don't want to allow to run an unlimited amount of queries in parallel. Field candidates are usually in the dozens, field value pairs can be in the hundreds.

Reply:
I don't see how running dozens of requests sequentially is a better experience. We can use pLimit here to limit the number of concurrent requests. Something like 5 sounds like a good start.

Reply:
+1 for using `pLimit` here

Reply:
`p-limit` sounds like a good idea worth pursuing but can we agree to do this in a follow up?

Reply:
yes that's fine 👍